npm/private/utils.bzl

```diff
@@ -156,6 +182,110 @@ def _npm_registry_download_url(package, version, registries, default_registry):
 def _is_git_repository_url(url):
     return url.startswith("git+ssh://")
 
+def _to_registry_url(url):
+    return "{}://{}".format(DEFAULT_REGISTRY_PROTOCOL, url) if url.find("//") == -1 else url
+
+def _default_registry():
+    return "{}://registry.npmjs.org/".format(DEFAULT_REGISTRY_PROTOCOL)
+
+def _hash(s):
+    result = str(hash(s))
+    if result.startswith("-"):
+        # Bazel's hash() resolves to a signed 32bit number [-2,147,483,648 to 2,147,483,647].
+        # Convert manually to a unique unsigned number here.
+        result = result[1:]
+        if len(result) <= 9:
+            padding = "0" * (9 - len(result))
+            result = "3" + padding + result
+        elif result[0] == "1":
+            result = "4" + result[1:]
```

Comment:
Why replace "1" with "4"? That makes this non-unique...?

Reply:
bash hash `function` returns a signed int `[-2,147,483,648 to 2,147,483,647]` which bothers me to put in to the hashes file.

So I did some string manipulation to make it a unique unsigned int string:
`-xxx,xxx,xxx` to `3,xxx,xxx,xxx`
`-1,xxx,xxx,xxx` to `4,xxx,xxx,xxx`
`-2,xxx,xxx,xxx` to `5,xxx,xxx,xxx`

Reply:
But now both `-1,xxx,xxx,xxx` and `4,xxx,xxx,xxx` result in the same hash (I think).

Shouldn't we prepend to make it unique instead of replacing the first char? Just replace any `-` with a `2` and put a `1` in front of non-negatives?

Reply:
`4,xxx,xxx,xxx` is out of the positive range so will never happen

Reply:
max positive int32 is 2,147,483,647

Reply:
Oh I see, that is quite easy to miss reading this.

If we're already making it too high to be an int I'd still vote to not trim it. Just put a `0` in front of a positive number and replace `-` with `1` in front of negative? 🤷

Reply:
sure. we can tweak in a follow-up.

Reply:
Normally that means it never happens :P

Also won't tweaking it in the future break caches?